compiler/optimizer/VPHandlers.cpp

```diff
@@ -1151,31 +1151,86 @@ TR::Node *constrainBCDAggrLoad(OMR::ValuePropagation *vp, TR::Node *node)
 TR::Node *constrainAnyIntLoad(OMR::ValuePropagation *vp, TR::Node *node)
    {
    TR::DataType dataType = node->getDataType();
-
    // Optimize characters being loaded out of the values array of a constant string
    //
-   if (dataType == TR::Int16 && node->getOpCode().isIndirect() &&
+   if (node->getOpCode().isIndirect() &&
        node->getSymbol()->isArrayShadowSymbol() &&
        node->getFirstChild()->getOpCode().isAdd())
       {
       TR::Node *array = node->getFirstChild()->getFirstChild();
       TR::Node *index = node->getFirstChild()->getSecondChild();
-      if (index->getOpCode().isLoadConst() && array->getOpCode().isIndirect())
+      if (index->getOpCode().isLoadConst())
          {
          bool isGlobal;
-         TR::Node* base = array->getFirstChild();
-         TR::VPConstraint *baseVPConstraint = vp->getConstraint(base, isGlobal);
-         if (baseVPConstraint && baseVPConstraint->isConstString())
+         if (dataType == TR::Int16 && array->getOpCode().isIndirect())
             {
-            TR::VPConstString *constString = baseVPConstraint->getClassType()->asConstString();
+            TR::Node* base = array->getFirstChild();
+            TR::VPConstraint *baseVPConstraint = vp->getConstraint(base, isGlobal);
+            if (baseVPConstraint && baseVPConstraint->isConstString())
+               {
+               TR::VPConstString *constString = baseVPConstraint->getClassType()->asConstString();
 
-            uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
-                                               - TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2, vp->comp());
-            if (ch != 0)
+               uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
```

Comment:
I guess getLongIntLow is the same as getLongInt for a constant index (as has been checked above) but it's still a slightly unexpected use of getLongIntLow here.

Reply:
```index``` here is really the "offset" (i.e. scaled by 2 value) is'nt it ? I assume the charAt api really expects the array offset to be passed in rather than the array index.

Reply:
Note that this line isn't really changing; it's just being reindented.

`index` (really the offset) is a constant node, not a constraint, and `getLongIntLow()` returns the low half of the 64-bit constant. A string can have a length up to 2^31-1 code units, or 2^32-2 bytes. Thankfully the valid offsets always fit in a 32-bit *unsigned* integer, i.e. there's no need to worry about losing the high half of a valid index (and in cases where the index is invalid, the load must be dead).

But this code takes the signed view, and offsets in [2^31,2^32-4] will be reinterpreted as negative. The negative offset will be converted to `uintptrj_t` before subtracting the array header size.

On 64-bit, we will sign-extend, then subtract the array header size, then do an unsigned 64-bit division by two, the result of which is still large enough that truncating to `int32_t` will produce a negative number, which is rejected by `charAt()`.

On 32-bit, we will reinterpret the signed offset as unsigned again, unsigned subtract the array header size (narrowly avoiding undefined behaviour in some cases), and do an unsigned division by 2, resulting in the correct index.

So I believe this code works, other than potentially missing some opportunities with very large offsets on 64-bit. However, it's extremely non-obvious, and that's why I suggested that this is not an example to emulate.

Reply:
Thanks for the explanation @jdmpapin 

Given the sub-optimality on 64-bit and the low likelihood of anyone analyzing this enough to change it, should we just make the minor change here to get the desired behaviour ?

Reply:
It would certainly be an improvement :+1:

Reply:
@jdmpapin I've updated with a change to fix this stuff up - could you buddy check it for me?